cmd/oras/internal/display/track/target.go

```diff
@@ -0,0 +1,121 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package track
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/registry"
+	"oras.land/oras/cmd/oras/internal/display"
+	"oras.land/oras/cmd/oras/internal/display/progress"
+)
+
+// Target tracks the progress of pushing ot underlying oras.Target.
+type Target struct {
+	oras.Target
+	manager      progress.Manager
+	actionPrompt string
+	donePrompt   string
+}
+
+// NewTarget creates a new tracked Target.
+func NewTarget(t oras.Target, actionPrompt, donePrompt string, tty *os.File) (*Target, error) {
+	manager, err := progress.NewManager(tty)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Target{
+		Target:       t,
+		manager:      manager,
+		actionPrompt: actionPrompt,
+		donePrompt:   donePrompt,
+	}, nil
+}
+
+// Push pushes the content to the Target with tracking.
+func (t *Target) Push(ctx context.Context, expected ocispec.Descriptor, content io.Reader) error {
+	r, err := managedReader(content, expected, t.manager, t.actionPrompt, t.donePrompt)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	r.Start()
+	if err := t.Target.Push(ctx, expected, r); err != nil {
+		return err
+	}
+	r.Done()
+	return nil
+}
+
+// PushReference pushes the content to the Target with tracking.
+func (t *Target) PushReference(ctx context.Context, expected ocispec.Descriptor, content io.Reader, reference string) error {
+	r, err := managedReader(content, expected, t.manager, t.actionPrompt, t.donePrompt)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	r.Start()
+	if rp, ok := t.Target.(registry.ReferencePusher); ok {
+		err = rp.PushReference(ctx, expected, r, reference)
+	} else {
+		if err := t.Target.Push(ctx, expected, r); err != nil {
+			return err
+		}
+		err = t.Target.Tag(ctx, expected, reference)
+	}
+	if err != nil {
+		return err
+	}
+	r.Done()
+	return nil
+}
+
+// Predecessors returns the predecessors of the node if supported.
+func (t *Target) Predecessors(ctx context.Context, node ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+	if p, ok := t.Target.(content.PredecessorFinder); ok {
+		return p.Predecessors(ctx, node)
+	}
+	return nil, fmt.Errorf("Target %v does not support Predecessors", reflect.TypeOf(t.Target))
+}
+
+// Close closes the tracking manager.
+func (t *Target) Close() error {
+	return t.manager.Close()
+}
+
+// Prompt prompts the user with the provided prompt and descriptor.
+// If Target is not set, only prints status.
+func (t *Target) Prompt(desc ocispec.Descriptor, prompt string, verbose bool) error {
+	if t == nil {
+		return display.PrintStatus(desc, prompt, verbose)
+	}
```

Comment:
Try to avoid the pattern of `t == nil`. It is hard to track and easy to have bugs.